p2p/metrics.go

```diff
@@ -19,53 +19,196 @@
+// system is disabled, the IP address is unspecified or the metered peer count
+// reached the limit, this function returns the original object.
+func newMeteredConn(conn net.Conn, ingress bool, ip net.IP) net.Conn {
 	// Short circuit if metrics are disabled
 	if !metrics.Enabled {
 		return conn
 	}
-	// Otherwise bump the connection counters and wrap the connection
+	if ip.IsUnspecified() {
+		log.Warn("Peer IP is unspecified")
+		return conn
+	}
+	if atomic.LoadUint64(&meteredPeerCount) >= MeteredPeerLimit {
+		log.Warn("Metered peer count reached the limit")
+		return conn
+	}
+	// Increment the metered peer count
+	atomic.AddUint64(&meteredPeerCount, 1)
+	// Bump the connection counters and wrap the connection
 	if ingress {
 		ingressConnectMeter.Mark(1)
 	} else {
 		egressConnectMeter.Mark(1)
 	}
-	return &meteredConn{Conn: conn}
+	return &meteredConn{
+		Conn:      conn,
+		ip:        ip,
+		connected: time.Now(),
+	}
 }
 
-// Read delegates a network read to the underlying connection, bumping the ingress
-// traffic meter along the way.
+// Read delegates a network read to the underlying connection, bumping the common
+// and the peer ingress traffic meters along the way.
 func (c *meteredConn) Read(b []byte) (n int, err error) {
 	n, err = c.Conn.Read(b)
 	ingressTrafficMeter.Mark(int64(n))
-	return
+	c.lock.RLock()
+	if c.ingressMeter != nil {
+		c.ingressMeter.Mark(int64(n))
+	}
+	c.lock.RUnlock()
+	return n, err
 }
 
-// Write delegates a network write to the underlying connection, bumping the
-// egress traffic meter along the way.
+// Write delegates a network write to the underlying connection, bumping the common
+// and the peer egress traffic meters along the way.
 func (c *meteredConn) Write(b []byte) (n int, err error) {
 	n, err = c.Conn.Write(b)
 	egressTrafficMeter.Mark(int64(n))
-	return
+	c.lock.RLock()
+	if c.egressMeter != nil {
+		c.egressMeter.Mark(int64(n))
+	}
+	c.lock.RUnlock()
+	return n, err
+}
+
+// handshakeDone is called when a peer handshake is done. Registers the peer to
+// the ingress and the egress traffic registries using the peer's IP and NodeID,
+// also emits connect event.
+func (c *meteredConn) handshakeDone(id discover.NodeID) {
+	c.lock.Lock()
+	c.id = id.String()
+	key := fmt.Sprintf("%s/%s", c.ip, c.id)
+	c.ingressMeter = metrics.NewRegisteredMeter(key, PeerIngressRegistry)
+	c.egressMeter = metrics.NewRegisteredMeter(key, PeerEgressRegistry)
+	c.lock.Unlock()
+
+	metricsFeed.Send(MeteredPeerEvent{
+		Type:    PeerConnected,
+		IP:      c.ip,
+		ID:      id.String(),
+		Elapsed: time.Since(c.connected),
+	})
+}
+
+// Close delegates a close operation to the underlying connection, unregisters
+// the peer from the traffic registries and emits close event.
+func (c *meteredConn) Close() error {
+	// Decrement the metered peer count
+	atomic.AddUint64(&meteredPeerCount, ^uint64(0))
+
+	c.lock.RLock()
+	// If the peer disconnects before the handshake
```

Comment:
Move this one live below into the `if` (`// Peer disconnected before the handshake`). Cleaner look.